xdsl/interactive/get_rewrites_and_passes.py

```diff
@@ -0,0 +1,98 @@
+from typing import NamedTuple
+
+from xdsl.dialects import builtin
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import PatternRewriter, RewritePattern
+from xdsl.tools.command_line_tool import get_all_dialects, get_all_passes
+from xdsl.transforms import individual_rewrite
+from xdsl.transforms.mlir_opt import MLIROptPass
+
+ALL_PASSES = tuple(sorted((p_name, p()) for (p_name, p) in get_all_passes().items()))
+"""Contains the list of xDSL passes."""
+
+
+class IndividualRewrite(NamedTuple):
+    """
+    Type alias for a possible rewrite, described by an operation and pattern name.
+    """
+
+    operation: str
+    pattern: str
+
+
+class IndexedIndividualRewrite(NamedTuple):
+    """
+    Type alias for a specific rewrite pattern, additionally consisting of its operation index.
+    """
+
+    operation_index: int
+    rewrite: IndividualRewrite
+
+
+def get_all_possible_rewrites(
+    patterns: tuple[IndividualRewrite, ...],
+    op: ModuleOp,
+    rewrite_by_name: dict[str, dict[str, RewritePattern]],
+) -> tuple[IndexedIndividualRewrite, ...]:
+    """
+    Function that takes a sequence of IndividualRewrite Patterns and a ModuleOp, and
+    returns the possible rewrites.
+    """
+    old_module = op.clone()
+    num_ops = len(list(old_module.walk()))
+
+    current_module = old_module.clone()
+
+    res: tuple[IndexedIndividualRewrite, ...] = ()
+
+    for op_idx in range(num_ops):
+        matched_op = list(current_module.walk())[op_idx]
+        if matched_op.name not in individual_rewrite.REWRITE_BY_NAMES:
+            continue
+        pattern_by_name = individual_rewrite.REWRITE_BY_NAMES[matched_op.name]
+
+        for pattern_name, pattern in pattern_by_name.items():
+            rewriter = PatternRewriter(matched_op)
+            pattern.match_and_rewrite(matched_op, rewriter)
+            if rewriter.has_done_action:
+                res = (
+                    *res,
+                    (
+                        IndexedIndividualRewrite(
+                            op_idx, IndividualRewrite(matched_op.name, pattern_name)
+                        )
+                    ),
+                )
+                current_module = old_module.clone()
+                matched_op = list(current_module.walk())[op_idx]
+
+    return res
+
+
+def condensed_pass_list(input: builtin.ModuleOp) -> tuple[type[ModulePass], ...]:
```

Comment:
why move this?

Reply:
I thought it might make sense to put the get rewrites and passes functions in one file to condense app.py. Should i move it back..?

Reply:
I'd recommend moving it back in this PR, even if it might make sense to move it to a different file in another future PR